File: 131-palindrome-partitioning/palindrome-partitioning.cpp
class Solution {
public:
    // Global
    vector<vector<string>> ans;
    vector<string> temp;

    void solve(string &s, int index){
        if (index >= s.length()) {
            ans.push_back(temp);
            return ;
        }

        for (int i=index; i<s.length(); i+=1){
            if (isPalindrome(s, index, i)){
                temp.push_back(s.substr(index, i-index+1));
                solve(s, i+1);
                temp.pop_back();
            }
        }
    }
    vector<vector<string>> partition(string s) {
        solve(s, 0);

        return ans;
    }

    bool isPalindrome(string &s, int st, int end){
        while (st <= end){
            if (s[st] != s[end]){
                return false;
            }
            st+=1;
            end-=1;
        }

        return true;
    }
};

File: 131-palindrome-partitioning/palindrome-partitioning.go
func solve(s string, idx int, temp []string, ans *[][]string) {
    if idx >= len(s) {
        // save the temp in the ans array
        tmp := make([]string, len(temp))
        copy(tmp, temp)
        *ans = append(*ans, tmp)

        return
    }

    for i := idx; i < len(s); i++ {
        if isPalindrome(s, idx, i) {
            temp = append(temp, s[idx:i+1])
            solve(s, i+1, temp, ans)
            temp = temp[:len(temp)-1]
        }
    }
}

func partition(s string) [][]string {
    var temp []string
    var ans [][]string
    solve(s, 0, temp, &ans)

    return ans
}

func isPalindrome(s string, st, end int) bool {
    for st < end {
        if s[st] != s[end] {
            return false
        }
        st++
        end--
    }

    return true
}